### PR TITLE
Generate variations from variation types

### PR DESCRIPTION
**Pull request by justinshreve**
This PR was a bit of a challenge because redux is new to me and took a bit of time to wrap my head around.

Variations are also tricky so I had to work-out the logic for generating new possible variations based on provided variation types while also taking into account existing variations (either from state or based on an API call). Basically implementing something similar to [this](https://github.com/woocommerce/woocommerce/blob/e4a6882eaf878428d1f098f8755cd915373f258a/includes/class-wc-ajax.php#L583) and [this](https://github.com/woocommerce/woocommerce/blob/71d0702523bcc365403dd3b84b945da5218d88b4/includes/wc-core-functions.php#L1051) in WooCommerce core.

I don't expect this PR to be perfect, but I think it's time to get some eyes/feedback on it so I can make some improvements. In terms of UI or front-facing changes, it'll generate rows in the variation table which will later house our settings like price (future PR). [Preview](https://cloudup.com/csgiqd9soGP).

So this PR:

* Hooks up the variation types form to variation generation so new variations can be created as new types are added/updated and new variations are made possible.
* Introduces state so that the types form can generate and pass variation data to the rest of the form.

To Test:

* Visit `products/add`
* Add a type named 'Color' with a 'Red' value.
* Add a type named 'Size' with a 'Small' value.
* Press "Create Variations".
* One variation should be created titled "Red - Small".
* Add a 'Blue' color.
* Press "Update Variations".
* Another variation should be created, titled "Blue - Small".
* Add a 'Medium' size.
* Press "Update Variations".
* Two new variations should be created, title "Red - Medium", and "Blue - Medium".
* Feel free to add additional types, try different labels/values, etc.

If needed, state can be cleared with: `localStorage.clear(); indexedDB.deleteDatabase( 'calypso' );`. The ability to remove variations will come in a later PR.

### CODE REVIEW

File: client/extensions/woocommerce/app/products/product-variation-types-form.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+				<div className="product-variation-types-form__group">
+					<div className="product-variation-types-form__labels">
+						<FormLabel className="product-variation-types-form__label">{ i18n.translate( 'Variation type' ) }</FormLabel>
+						<FormLabel>{ i18n.translate( 'Variation values' ) }</FormLabel>
+					</div>
+					{inputs}
+				</div>
+
+				<Button onClick={ this.addVariationType }>{ i18n.translate( 'Add another variation type' ) }</Button>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-variation-types-form.js]: Hi! I've found a possible matching string that has already been translated **1** time:
`translate( 'Add another variation' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by coderkevin** [comment]
This looks like a good start here, but I think we need some dialog about what precedents we want to set on this kind of code. I'd like to review this interactively when you have the chance, just ping me and let me know when you're available. Thanks!

File: client/extensions/woocommerce/app/products/product-variation-types-form.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		const existingVariations = [];
+		const variations = [];
+
+		/*
+		 * Create an array of each set of possible values for a specific type
+		 * so that we can generate a new cartesian of products.
+		 * We also need to store a relation of indexes to the type label so we can
+		 * properly build new variation objects later.
+		*/
+		variationTypes.map( function( variationType ) {
```
Comment (blowery) [inline client/extensions/woocommerce/app/products/product-variation-types-form.js]: could use `forEach` instead of `map` since you're only iterating for side-effects, not actually using the return.
